src/ConstructedStyleSheet.ts

```diff
@@ -0,0 +1,240 @@
+  if (!$basicStyleSheet.has(self)) {
+    throw new TypeError('Illegal invocation');
+  }
+}
+
+/**
+ * A replacement for CSSStyleSheet class which is not actually constructable in
+ * any browser that does not support Constructable Style Sheet proposal. To
+ * fulfil the Constructable Style Sheet specification, it preserves a separate
+ * HTMLStyleElement that behaves as a backup for all the `<style>` adopters.
+ */
+declare class ConstructedStyleSheet extends CSSStyleSheet {
+  replace(text: string): Promise<ConstructedStyleSheet>;
+  replaceSync(text: string): void;
+}
+
+function ConstructedStyleSheet(this: ConstructedStyleSheet) {
+  var style = document.createElement('style');
+  bootstrapper.body.appendChild(style);
+
+  // Init private properties
+  $basicStyleSheet.set(this, style.sheet!);
+  $locations.set(this, []);
+  $adoptersByLocation.set(this, new WeakMap());
+}
+
+var proto = ConstructedStyleSheet.prototype;
+
+proto.replace = function replace(
+  contents: string,
+): Promise<ConstructedStyleSheet> {
+  try {
+    this.replaceSync(contents);
+
+    return Promise.resolve(this);
+  } catch (e) {
+    return Promise.reject(e);
+  }
+};
+
+proto.replaceSync = function replaceSync(contents) {
+  // CSSStyleSheet.prototype.replaceSync('body {}')
+  checkInvocationCorrectness(this);
+
+  if (typeof contents === 'string') {
+    var self = this;
+
+    var basic = $basicStyleSheet.get(self)!;
+    var sanitized = rejectImports(contents);
+    clearRules(basic);
+
+    if (sanitized) {
+      basic.insertRule(sanitized, 0);
+    }
+
+    $locations.get(self)!.forEach(function (location) {
+      if (location.isConnected()) {
+        // Type Note: if location is connected, adopter is already created.
+        restyleAdopter(self, getAdopterByLocation(self, location)!);
+      }
+    });
+  }
+};
+
+defineProperty(proto, 'cssRules', {
+  configurable: true,
+  enumerable: true,
+  get: function cssRules() {
+    // CSSStyleSheet.prototype.cssRules;
+    checkInvocationCorrectness(this);
+
+    return $basicStyleSheet.get(this)!.cssRules;
+  },
+});
+
+cssStyleSheetMethods.forEach(function (method) {
+  proto[method] = function () {
+    var self = this;
+    checkInvocationCorrectness(self);
+
+    var args = arguments;
+    var basic = $basicStyleSheet.get(self)!;
+    var locations = $locations.get(self)!;
+
+    var result = basic[method].apply(basic, args);
+
+    locations.forEach(function (location) {
+      if (location.isConnected()) {
+        // Type Note: If location is connected, adopter is already created; and
+        // since it is connected to DOM, the sheet cannot be null.
+        var sheet = getAdopterByLocation(self, location)!.sheet!;
+        sheet[method].apply(sheet, args);
+      }
+    });
+
+    return result;
+  };
+});
+
+defineProperty(ConstructedStyleSheet, Symbol.hasInstance, {
+  configurable: true,
+  value: isCSSStyleSheetInstance,
+});
```

Comment:
Given that the logged version of the constructed stylesheet is just `{}`, I'm wondering if we shouldn't have a `Symbol.toTagString` or some other thing here as well. Not a huge deal, just thinking aloud.

Reply:
Not working unfortunately. It works only if you do:
```js
var a = new CSSStyleSheet();
console.log(a.toString());
```

I suggest omitting that part because it is unnecessary for polyfill but adds extra bytes.